### PR TITLE
Improve unix command history plugin

### PR DESCRIPTION
**Pull request by JSCU-CNI**
This patch makes the bashhistory unix function shell-agnostic.

### CODE REVIEW

**Comment by pyrco** [comment]
Hi, thanks for the nice extension of the commandhistory plugin!

Before it can be merged in however, some changes need to be made to the `RecordDescriptorExtention` used for the `_user` parameter of the `CommandHistoryRecord`.

The design/idea behind this extension field is that it is uniform across different OSes and lightweight on information to prevent to much data duplication in the output. As you can imagine the latter is of course a balance, hence the inclusion of e.g. `user_home`. Any other information can be enriched by correlating these records with the output of the users plugin. So this means using a `UnixUserRecordDescriptorExtension` for the `_user` parameter is unfortunately not what should be done.

We discussed internally and agreed that adding a `user_group` field to the `UserRecordDescriptorExtension` class in `helpers/descripter_extensions.py` which contains the id of the main group of the user should not extend the `UserRecordDescriptorExtension`  too much. So please implement it there. For now it is ok to leave the value for `user_group` as `None` on windows based systems.

File: dissect/target/helpers/descriptor_extensions.py [hunk 1]
```diff
@@ -81,3 +81,25 @@ def _fill_default_fields(self, record_kwargs):
         # Reserved keywords are never part of the args of a Record's __init__.
         record_kwargs["_source"] = source
         return record_kwargs
+
+
+class UnixUserRecordDescriptorExtension(RecordDescriptorExtensionBase):
```
Comment (pyrco) [inline dissect/target/helpers/descriptor_extensions.py:86]: See earlier comment

**Review by pyrco — CHANGES_REQUESTED** [review]

**Comment by JSCU-CNI** [comment]
Thank you for your comment @pyrco. I added `user_group` to the generic `UserRecordDescriptorExtension`.

Could you clarify why `WindowsUserRecord` and `UnixUserRecord` exist in `dissect/target/helpers/record.py`? It seems to me that per your rhetoric they should be merged into a single `UserRecordDescriptorExtension`.

File: dissect/target/plugins/os/unix/commandhistory.py [hunk 2]
```diff
@@ -5,34 +5,54 @@
...
+                    continue
+
             for file_ in user_details.home_path.iterdir():
-                if not file_.name.startswith(".bash_history"):
+                if file_.name not in COMMAND_HISTORY_FILES:
                     continue
 
                 try:
-                    for line in file_.open("rt"):
+                    for line in file_.open("rt", errors="ignore"):  # Ignore Non-ASCII characters in bash_history
```
Comment (pyrco) [inline dissect/target/plugins/os/unix/commandhistory.py:55]:
```suggestion
                    for line in file_.open("rt", errors="ignore"):  # Ignore Non-UTF-8 characters in bash_history
```

Wouldn't it be better to use `errors="replace"` so you know you are potentially missing something?

File: dissect/target/plugins/os/unix/commandhistory.py [hunk 3]
```diff
@@ -5,34 +5,54 @@
...
+        return False
 
-    @export(record=BashHistoryRecord)
+    @export(record=CommandHistoryRecord)
     def bashhistory(self):
-        """Return bash history for all users.
+        """
+        Deprecated, use commandhistory function.
+        """
+        return self.commandhistory()
```
Comment (pyrco) [inline dissect/target/plugins/os/unix/commandhistory.py:34]:
```suggestion
        warnings.warn("bashhistory is deprecated, use the commandhistory function", DeprecationWarning)
        return self.commandhistory()
```

**Review by pyrco — CHANGES_REQUESTED** [review]

**Comment by JSCU-CNI** [comment]
We added proper support for timestamps in bash history files per the `HISTTIMEFORMAT` setting (https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#index-HISTTIMEFORMAT). Support for Zsh timestamps is not yet implemented.

**Comment by pyrco** [comment]
I saw it, nice! It is much cleaner this way.

**Review by pyrco — APPROVED** [review]

**Review by pyrco — CHANGES_REQUESTED** [review]
_Could you rebase this patch set on the current main branch? There appear to be some conflicting changes. Once rebased, I'll merge it in._

Nevermind, I, merged main cleanly into the branch.

**Review by pyrco — APPROVED** [review]

**Pull request merged**